File: src/server/spatialGrid.ts
import { CreatureAgent, FoodSpore } from "../shared/types";

export class SpatialGrid {
  public readonly cellSize = 80;
  public readonly cols = 240; // 19200 / 80
  public readonly rows = 135; // 10800 / 80

  private creatureBuckets: CreatureAgent[][];
  private foodBuckets: FoodSpore[][];

  constructor() {
    const totalCells = this.cols * this.rows;
    this.creatureBuckets = Array.from({ length: totalCells }, () => []);
    this.foodBuckets = Array.from({ length: totalCells }, () => []);
  }

  public clear() {
    const totalCells = this.cols * this.rows;
    for (let i = 0; i < totalCells; i++) {
      this.creatureBuckets[i].length = 0;
      this.foodBuckets[i].length = 0;
    }
  }

  private getCellIndex(x: number, y: number): number {
    const c = Math.max(0, Math.min(this.cols - 1, Math.floor(x / this.cellSize)));
    const r = Math.max(0, Math.min(this.rows - 1, Math.floor(y / this.cellSize)));
    return c * this.rows + r;
  }

  public insertCreature(agent: CreatureAgent) {
    const idx = this.getCellIndex(agent.px, agent.py);
    this.creatureBuckets[idx].push(agent);
  }

  public insertFood(pellet: FoodSpore) {
    const idx = this.getCellIndex(pellet.x, pellet.y);
    this.foodBuckets[idx].push(pellet);
  }

  /**
   * Returns all creatures within a bounding box centered at (x, y) with a given range.
   */
  public getNearbyCreatures(x: number, y: number, range: number): CreatureAgent[] {
    const startCol = Math.max(0, Math.floor((x - range) / this.cellSize));
    const endCol = Math.min(this.cols - 1, Math.floor((x + range) / this.cellSize));
    const startRow = Math.max(0, Math.floor((y - range) / this.cellSize));
    const endRow = Math.min(this.rows - 1, Math.floor((y + range) / this.cellSize));

    const result: CreatureAgent[] = [];
    for (let c = startCol; c <= endCol; c++) {
      for (let r = startRow; r <= endRow; r++) {
        const idx = c * this.rows + r;
        const bucket = this.creatureBuckets[idx];
        for (let i = 0; i < bucket.length; i++) {
          result.push(bucket[i]);
        }
      }
    }
    return result;
  }

  /**
   * Returns all food pellets within a bounding box centered at (x, y) with a given range.
   */
  public getNearbyFood(x: number, y: number, range: number): FoodSpore[] {
    const startCol = Math.max(0, Math.floor((x - range) / this.cellSize));
    const endCol = Math.min(this.cols - 1, Math.floor((x + range) / this.cellSize));
    const startRow = Math.max(0, Math.floor((y - range) / this.cellSize));
    const endRow = Math.min(this.rows - 1, Math.floor((y + range) / this.cellSize));

    const result: FoodSpore[] = [];
    for (let c = startCol; c <= endCol; c++) {
      for (let r = startRow; r <= endRow; r++) {
        const idx = c * this.rows + r;
        const bucket = this.foodBuckets[idx];
        for (let i = 0; i < bucket.length; i++) {
          result.push(bucket[i]);
        }
      }
    }
    return result;
  }
}

File: src-tauri/src/shared/spatial_grid.rs
pub struct SpatialGrid {
    pub cell_size: f32,
    pub cols: usize,
    pub rows: usize,
    creature_buckets: Vec<Vec<u32>>,
    food_buckets: Vec<Vec<u32>>,
}

impl SpatialGrid {
    pub fn new(width: f32, height: f32, cell_size: f32) -> Self {
        let cols = (width / cell_size).ceil() as usize;
        let rows = (height / cell_size).ceil() as usize;
        let total_cells = cols * rows;

        Self {
            cell_size,
            cols,
            rows,
            creature_buckets: vec![Vec::new(); total_cells],
            food_buckets: vec![Vec::new(); total_cells],
        }
    }

    pub fn clear(&mut self) {
        for bucket in &mut self.creature_buckets {
            bucket.clear();
        }
        for bucket in &mut self.food_buckets {
            bucket.clear();
        }
    }

    fn get_cell_index(&self, x: f32, y: f32) -> usize {
        let c = ((x / self.cell_size).floor() as usize).clamp(0, self.cols.saturating_sub(1));
        let r = ((y / self.cell_size).floor() as usize).clamp(0, self.rows.saturating_sub(1));
        c * self.rows + r
    }

    pub fn insert_creature(&mut self, id: u32, x: f32, y: f32) {
        let idx = self.get_cell_index(x, y);
        if idx < self.creature_buckets.len() {
            self.creature_buckets[idx].push(id);
        }
    }

    pub fn insert_food(&mut self, id: u32, x: f32, y: f32) {
        let idx = self.get_cell_index(x, y);
        if idx < self.food_buckets.len() {
            self.food_buckets[idx].push(id);
        }
    }

    pub fn get_nearby_creatures(&self, x: f32, y: f32, range: f32) -> Vec<u32> {
        let start_col = ((x - range) / self.cell_size).floor() as usize;
        let end_col = ((x + range) / self.cell_size).floor() as usize;
        let start_row = ((y - range) / self.cell_size).floor() as usize;
        let end_row = ((y + range) / self.cell_size).floor() as usize;

        let start_col = start_col.clamp(0, self.cols.saturating_sub(1));
        let end_col = end_col.clamp(0, self.cols.saturating_sub(1));
        let start_row = start_row.clamp(0, self.rows.saturating_sub(1));
        let end_row = end_row.clamp(0, self.rows.saturating_sub(1));

        let mut result = Vec::new();
        for c in start_col..=end_col {
            for r in start_row..=end_row {
                let idx = c * self.rows + r;
                if idx < self.creature_buckets.len() {
                    result.extend(&self.creature_buckets[idx]);
                }
            }
        }
        result
    }

    pub fn get_nearby_food(&self, x: f32, y: f32, range: f32) -> Vec<u32> {
        let start_col = ((x - range) / self.cell_size).floor() as usize;
        let end_col = ((x + range) / self.cell_size).floor() as usize;
        let start_row = ((y - range) / self.cell_size).floor() as usize;
        let end_row = ((y + range) / self.cell_size).floor() as usize;

        let start_col = start_col.clamp(0, self.cols.saturating_sub(1));
        let end_col = end_col.clamp(0, self.cols.saturating_sub(1));
        let start_row = start_row.clamp(0, self.rows.saturating_sub(1));
        let end_row = end_row.clamp(0, self.rows.saturating_sub(1));

        let mut result = Vec::new();
        for c in start_col..=end_col {
            for r in start_row..=end_row {
                let idx = c * self.rows + r;
                if idx < self.food_buckets.len() {
                    result.extend(&self.food_buckets[idx]);
                }
            }
        }
        result
    }
}
